### PR TITLE
feat(bitbucket-server): Support gitUrl

### PR DESCRIPTION
**Pull request by jatladams**
## Changes:

This allows the gitUrl property to be used in bitbucket-server to set how repositories should be accessed.

If no value is provided for gitUrl the previous logic will be used. This should mean users who upgrade without change see no difference in functionality.

If a gitUrl value is provided as per https://docs.renovatebot.com/self-hosted-configuration/#giturl:
default: use HTTPS URLs provided by the platform for Git or return undefined if none provided
ssh: use SSH URLs provided by the platform for Git or return undefined if none provided
endpoint: ignore URLs provided by the platform and use the configured endpoint directly

Added tests to 'utils.spec.ts' to cover 'getRepoGitUrl'
Added tests to 'index.spec.ts' to cover 'gitUrl' values

## Context:

Currently there is no way to specify a specific git connection type. Because Bitbucket Server provides all available connection types in the [API response ](https://github.com/renovatebot/renovate/blob/9e804e9cb7e9ed12fb66f031dd4e611cc8726efb/lib/modules/platform/bitbucket-server/index.ts#L179)  (http and ssh) the only way to force ssh connection currently is to turn off http connection.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs docs added, but first please create an issue to discuss your requirements and proposed implementation approach

**Comment by jatladams** [comment]
@rarkins There is an existing issue for this change: https://github.com/renovatebot/renovate/issues/9607
Is there a way to attach this change to that issue?

**Comment by rarkins** [comment]
You can type `Closes #` in the Context of PRs to link the two. However that issue has label `status:requirements` which means it needs further discussion before confirming it's ready for implementation

**Comment by jatladams** [comment]
Ah I missed that tag, my mistake.

I saw the issue and just followed the implementation in `https://github.com/renovatebot/renovate/blob/main/lib/platform/gitlab/index.ts getRepoUrl()`

Are there any steps I can take to move the issue forward?

**Comment by viceice** [comment]
Please do the refactoring first in a separate PR, so we can more easily review.

**Comment by jatladams** [comment]
Is this what you had in mind @viceice ?
https://github.com/renovatebot/renovate/commit/abbe1773b59a4b8a075bac051307780a05b6af2b?diff=split

**Comment by viceice** [comment]
Yes, the getRepoUrl function extraction

**Comment by harmjanblok** [comment]
Seems already good progress has been made on this issue. What are the remaining open points for this to be merged?

**Review by viceice — CHANGES_REQUESTED** [review]
needs deconflicting

👉 https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

File: lib/platform/bitbucket-server/index.ts [hunk 1]
```diff
@@ -153,11 +159,60 @@ export async function getJsonFile(
   return JSON.parse(raw);
 }
 
+function getRepoUrl(
```
Comment (viceice) [inline lib/platform/bitbucket-server/index.ts:162]: Please do this function extraction first in a separate PR without any functional change. Then use this branch / PR for the  gitUrl` changes only.

**Comment by jatladams** [comment]
Apologies I haven't had a chance to fix up the [refactor](https://github.com/jatladams/renovate/tree/bbs-index-refactor) requested to move this forward. I will try find some time this week to work on it

**Comment by rarkins** [comment]
Needs deconflicting

**Comment by jatladams** [comment]
Hi @rarkins and @viceice,
I have resolved the merge conflicts, updated the logic so that users will so no change unless they specify a value for gitUrl and added unit tests to ensure coverage of the new code.

For the new functionality I have matched the logic as specified in https://docs.renovatebot.com/self-hosted-configuration/#giturl or return undefined if the expected values are not returned from Bitbucket Server API

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 2]
```diff
@@ -291,6 +291,117 @@ describe('modules/platform/bitbucket-server/index', () => {
           expect(httpMock.getTrace()).toMatchSnapshot();
         });
 
+        it('no git url', async () => {
```
Comment (jatladams) [inline lib/modules/platform/bitbucket-server/index.spec.ts:294]: I think this test may not have been correctly resolved in the merge, I will investigate

File: lib/modules/platform/bitbucket-server/index.ts [hunk 3]
```diff
@@ -191,16 +192,17 @@ export async function initRepo({
...
       defaults.endpoint,
+      gitUrl,
       info,
       opts
     );
 
     await git.initRepo({
       ...config,
-      url: gitUrl,
+      url: url,
```
Comment (rarkins) [inline lib/modules/platform/bitbucket-server/index.ts:205]:
```suggestion
      url,
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Please also update the description inside docs/usage/

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 4]
```diff
@@ -291,6 +291,117 @@ describe('modules/platform/bitbucket-server/index', () => {
...
+            .reply(200, {
+              displayId: 'master',
+            });
+          expect(
+            await bitbucket.initRepo({
+              endpoint: 'https://stash.renovatebot.com/vcs/',
+              repository: 'SOME/repo',
+            })
+          ).toMatchSnapshot();
+          expect(httpMock.getTrace()).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:312]: remove

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 5]
```diff
@@ -291,6 +291,117 @@ describe('modules/platform/bitbucket-server/index', () => {
...
+            )
+            .reply(200, {
+              displayId: 'master',
+            });
+          expect(
+            await bitbucket.initRepo({
+              endpoint: 'https://stash.renovatebot.com/vcs/',
+              repository: 'SOME/repo',
+            })
+          ).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:311]: use `toEqual` matcher

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 6]
```diff
@@ -291,6 +291,117 @@ describe('modules/platform/bitbucket-server/index', () => {
...
+          const res = await bitbucket.initRepo({
+            endpoint: 'https://stash.renovatebot.com/vcs/',
+            repository: 'SOME/repo',
+            gitUrl: 'ssh',
+          });
+          expect(git.initRepo).toHaveBeenCalledWith(
+            expect.objectContaining({ url: sshLink('SOME', 'repo') })
+          );
+          expect(res).toMatchSnapshot();
+          expect(httpMock.getTrace()).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:339]: same here. please also fix all other new/ changed tests.

File: lib/modules/platform/bitbucket-server/utils.spec.ts [hunk 7]
```diff
@@ -1,11 +1,70 @@
...
+            name: 'http',
+          }
+        : null,
+      options.cloneUrl.ssh
+        ? {
+            href: sshLink(projectKey, repositorySlug),
+            name: 'ssh',
+          }
+        : null,
+    ].filter(Boolean);
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.spec.ts:62]: use `is.truthy`

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 8]
```diff
@@ -155,38 +157,80 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
 }
 
+function generateUrlFromEndpoint(
+  defaultEndpoint: string,
+  opts: HostRule,
+  repository: string
+): string {
+  const { host, pathname } = url.parse(defaultEndpoint);
+  const generatedUrl = git.getUrl({
+    protocol: defaultEndpoint.split(':')[0] as GitProtocol,
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:167]: why not using from parsed url?

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 9]
```diff
@@ -155,38 +157,80 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
+      } else {
+        logger.warn(
+          `endpoint URL could not be found for ${repository}. Falling back to generating`
+        );
+      }
+    } else if (gitUrl === 'endpoint') {
+      // Fallback to generating the url if the API didn't give us an URL
+      return generateUrlFromEndpoint(defaultEndpoint, opts, repository);
+    }
+    return undefined;
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:234]: use `null` instead of `undefined`

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 10]
```diff
@@ -155,38 +157,80 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
+  return generatedUrl;
+}
+
 export function getRepoGitUrl(
   repository: string,
   defaultEndpoint: string,
+  gitUrl: GitUrlOption,
   info: BbsRestRepo,
   opts: HostRule
 ): string {
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:184]:
```suggestion
 ): string | null {
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket-server/utils.ts (hunk 8 above)
Comment (jatladams) [reply lib/modules/platform/bitbucket-server/utils.ts]: Good point, I also changed the deprecated `url.parse(defaultEndpoint);` for `new URL(defaultEndpoint);`

**Comment by jatladams** [comment]
I have applied the feedback, I still need to update the docs
- [ ] update docs/usage/

File: lib/modules/platform/bitbucket-server/utils.ts (hunk 8 above)
Comment (viceice) [reply lib/modules/platform/bitbucket-server/utils.ts]: use `parseUrl` from `util/url`

Comment (viceice) [reply lib/modules/platform/bitbucket-server/utils.ts]: and please do the refactoring / preparation in a separate PR to be merged first

**Comment by rarkins** [comment]
Needs deconflicting

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 11]
```diff
@@ -155,39 +157,81 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
+    repository,
+  });
+  logger.debug({ url: generatedUrl }, `using generated endpoint URL`);
+  return generatedUrl;
+}
+
 export function getRepoGitUrl(
   repository: string,
   defaultEndpoint: string,
+  gitUrl: GitUrlOption | undefined,
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:182]: this function can be simplified, use a switch case on `gitUrl` and throw a config error if `gitUrl` is set, but type couldn't found ( http, ssh )

**Review by viceice — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md [hunk 12]
```diff
@@ -370,7 +370,7 @@ To handle the case where the underlying Git processes appear to hang, configure
 ## gitUrl
 
 Override the default resolution for Git remote, e.g. to switch GitLab from HTTPS to SSH-based.
-Currently works for GitLab only.
+Currently works for Bitbucket Server and GitLab only.
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:373]:
```suggestion
```

Add `supportedPlatform` to its config options definition instead

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md (hunk 12 above)
Comment (jatladams) [reply docs/usage/self-hosted-configuration.md:373]: Is there another config option to update besides https://github.com/renovatebot/renovate/pull/13875/commits/affda143098433a19f6d249d01ac0cf21528fd3c ?

File: lib/modules/platform/bitbucket-server/utils.ts (hunk 11 above)
Comment (jatladams) [reply lib/modules/platform/bitbucket-server/utils.ts:182]: Do you mean something like:
`  switch(gitUrl) {
    case "default": {
      // do something;
      break;
    }
    case "ssh": {
      // do something;
      break;
    }
    case "endpoint": {
      // do something;
      break;
    }
    case undefined: {
      // do something;
      break;
    }
    default: {
      //throw error;
      break;
    }
  }`

**Comment by mintmac** [comment]
I'm also after this feature. Any chances to push this forward?

**Comment by viceice** [comment]
needs deconfliction

**Comment by submac** [comment]
I have a very little knowledge of js/ts. Could @jatladams or someone else help to close it?

File: docs/usage/self-hosted-configuration.md (hunk 12 above)
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:386]: No other

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 13]
```diff
@@ -156,39 +158,81 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
   info: BbsRestRepo,
   opts: HostRule
-): string {
-  let cloneUrl = info.links.clone?.find(({ name }) => name === 'http');
-  if (!cloneUrl) {
-    // Http access might be disabled, try to find ssh url in this case
-    cloneUrl = info.links.clone?.find(({ name }) => name === 'ssh');
-  }
+): string | null {
+  if (gitUrl === undefined) {
```
Comment (rarkins) [inline lib/modules/platform/bitbucket-server/utils.ts:187]: @viceice do the latest typing changes mean that this is now not possible - would be `null`?

**Comment by jatladams** [comment]
@submac I can look at the `utils.ts` that are requested and the lint failures.

File: lib/modules/platform/bitbucket-server/utils.ts (hunk 11 above)
Comment (jatladams) [reply lib/modules/platform/bitbucket-server/utils.ts:183]: Changes applied. 

I used undefined to follow the existing logic where we try http then ssh and finally generate the URL.

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 14]
```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
+      const sshUrl = getRepoGitUrl(
+        repository,
+        defaultEndpoint,
+        'ssh',
+        info,
+        opts
+      );
+      if (sshUrl !== null) {
+        return sshUrl;
+      }
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:240]:
```suggestion
): string {
  if (gitUrl === 'ssh') {
    const sshUrl = info.links.clone?.find(({ name }) => name === 'ssh');
    if (sshUrl === undefined) {
      throw new Error(CONFIG_GIT_URL_UNAVAILABLE);
    }

    logger.debug({ url: sshUrl.href }, `using ssh URL`);
    return sshUrl.href;
  }

  let cloneUrl = info.links.clone?.find(({ name }) => name === 'http');
  if (cloneUrl) {
    // Inject auth into the API provided URL
    return injectAuth(cloneUrl.href, opts);
  }

  // Http access might be disabled, try to find ssh url in this case
  cloneUrl = info.links.clone?.find(({ name }) => name === 'ssh');

  if (gitUrl === 'endpoint' || !cloneUrl) {
    return generateUrlFromEndpoint(defaultEndpoint, opts, repository);
  }

  // SSH urls can be used directly
  return cloneUrl.href;
}
```

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 15]
```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
+    auth: `${opts.username}:${opts.password}`,
+    host: `${url.host}${url.pathname}${
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+      url.pathname!.endsWith('/') ? '' : /* istanbul ignore next */ '/'
+    }scm`,
+    repository,
+  });
+  logger.debug({ url: generatedUrl }, `using generated endpoint URL`);
+  return generatedUrl;
+}
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:178]:
```suggestion
}

function injectAuth(url: string, opts: HostRule): string {
  const repoUrl = parseUrl(url)!;
  if (!repoUrl) {
    logger.debug(`Invalid url: ${url}`);
    throw new Error(CONFIG_GIT_URL_UNAVAILABLE);
  }
  // TODO: null checks (#7154)
  repoUrl.username = opts.username!;
  repoUrl.password = opts.password!;
  return repoUrl.toString();
}
```

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 16]
```diff
@@ -11,6 +12,7 @@ import type {
   HttpResponse,
 } from '../../../util/http/types';
 import { getPrBodyStruct } from '../pr-body';
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:14]:
```suggestion
import { ensureTrailingSlash, parseUrl } from '../../../util/url';
import { getPrBodyStruct } from '../pr-body';
```

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 17]
```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
-    // SSH urls can be used directly
-    gitUrl = cloneUrl.href;
+      return getRepoGitUrl(repository, defaultEndpoint, 'endpoint', info, opts);
+    }
+    default: {
+      throw new TypeError(
+        `The specified scm connection type ${gitUrl} is not a valid GitUrlOption value`
+      );
+    }
   }
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:249]:
```suggestion
```

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 18]
```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
...
-  return gitUrl;
 }
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:250]:
```suggestion
```

**Review by viceice — COMMENTED** [review]
simplify things, sorry switch made it even more complex

File: lib/modules/platform/bitbucket-server/index.ts [hunk 19]
```diff
@@ -189,17 +191,22 @@ export async function initRepo({
...
       defaults.endpoint!,
+      gitUrl,
       info,
       opts
     );
 
+    if (url === null) {
+      throw new Error(CONFIG_GIT_URL_UNAVAILABLE);
+    }
+
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:206]:
```suggestion
```

File: lib/modules/platform/bitbucket-server/index.ts [hunk 20]
```diff
@@ -4,6 +4,7 @@ import JSON5 from 'json5';
 import type { PartialDeep } from 'type-fest';
 import { PlatformId } from '../../../constants';
 import {
+  CONFIG_GIT_URL_UNAVAILABLE,
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:7]:
```suggestion
```

File: lib/modules/platform/bitbucket-server/utils.ts [hunk 21]
```diff
@@ -1,6 +1,7 @@
 // SEE for the reference https://github.com/renovatebot/renovate/blob/c3e9e572b225085448d94aa121c7ec81c14d3955/lib/platform/bitbucket/utils.js
 import url from 'url';
 import is from '@sindresorhus/is';
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/utils.ts:3]:
```suggestion
import is from '@sindresorhus/is';
import { CONFIG_GIT_URL_UNAVAILABLE } from '../../../constants/error-messages';
```

File: lib/modules/platform/bitbucket-server/utils.ts (hunk 16 above)
Comment (jatladams) [reply lib/modules/platform/bitbucket-server/utils.ts:14]: `ensureTrailingSlash` doesn't appear to be used, is it missing from another change?

File: lib/modules/platform/bitbucket-server/utils.ts (hunk 14 above)
Comment (jatladams) [reply lib/modules/platform/bitbucket-server/utils.ts:240]: I am slightly concerned a user could set `gitUrl` to `endpoint` and it would be ignored if `http` or `ssh` are available. Is this the desired functionality?

File: lib/modules/platform/bitbucket-server/utils.ts (hunk 16 above)
Comment (viceice) [reply lib/modules/platform/bitbucket-server/utils.ts:16]: yes, sorry, leftover from testing

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]
